base/multidimensional.jl

```diff
@@ -440,44 +492,78 @@ module IteratorsMD
-    @inline function iterate(r::Reverse{<:CartesianIndices}, state)
-        valid, I = __dec(state.I, last(r.itr).I, first(r.itr).I)
+    @propagate_inbounds function iterate(r::Reverse{<:CartesianIndices}, state)
+        valid, I = __dec(state.I, r.itr.indices)
         valid || return nothing
         return CartesianIndex(I...), CartesianIndex(I...)
     end
 
     # decrement & carry
-    @inline function dec(state, start, stop)
-        _, I = __dec(state, start, stop)
+    @propagate_inbounds function dec(state, indices)
+        _, I = __dec(state, indices)
         return CartesianIndex(I...)
     end
 
     # decrement post check to avoid integer overflow
-    @inline __dec(::Tuple{}, ::Tuple{}, ::Tuple{}) = false, ()
-    @inline function __dec(state::Tuple{Int}, start::Tuple{Int}, stop::Tuple{Int})
-        valid = state[1] > stop[1]
-        return valid, (state[1]-1,)
+    @propagate_inbounds __dec(::Tuple{}, ::Tuple{}) = false, ()
+    @propagate_inbounds function __dec(state::Tuple{Int}, indices::Tuple{<:OrdinalRange})
+        rng = indices[1]
+        I = state[1] - step(rng)
+        valid = __isvalid_dec_range(I, rng) && __has_dec_overflow(I, rng)
+        return valid, (I,)
     end
 
-    @inline function __dec(state, start, stop)
-        if state[1] > stop[1]
-            return true, (state[1]-1, tail(state)...)
+    @propagate_inbounds function __dec(state, indices)
+        rng = indices[1]
+        I = state[1] - step(rng)
+        if __isvalid_dec_range(I, rng) && __has_dec_overflow(I, rng)
+            return true, (I, tail(state)...)
         end
-        valid, I = __dec(tail(state), tail(start), tail(stop))
-        return valid, (start[1], I...)
+        valid, I = __dec(tail(state), tail(indices))
+        return valid, (last(rng), I...)
+    end
+
+    @inline __isvalid_dec_range(I, rng) = __isvalid_range(I, rng)
+    @inline __isvalid_dec_range(I, rng::AbstractUnitRange) = I >= first(rng)
+
+    @propagate_inbounds __has_dec_overflow(I, rng) = __has_overflow(I, rng)
+    @inline function __has_dec_overflow(I::T, rng::AbstractUnitRange) where T
+        @boundscheck return I <= last(rng)
+        return true
     end
 
     # 0-d cartesian ranges are special-cased to iterate once and only once
     iterate(iter::Reverse{<:CartesianIndices{0}}, state=false) = state ? nothing : (CartesianIndex(), true)
 
-    Base.LinearIndices(inds::CartesianIndices{N,R}) where {N,R} = LinearIndices{N,R}(inds.indices)
+    Base.LinearIndices(inds::CartesianIndices{N,R}) where {N,R<:AbstractUnitRange} = LinearIndices{N,R}(inds.indices)
+    function Base.LinearIndices(inds::CartesianIndices{N}) where N
+        indices = inds.indices
+        if all(x->x==1, step.(indices))
+            indices = map(ind->convert(AbstractUnitRange, ind), indices)
+            LinearIndices{N, typeof(indices)}(indices)
+        else
+            throw(ArgumentError("LinearIndices for $typeof(inds) with non-1 step size is not supported."))
```

Comment:
Actually, thinking about this more, maybe we can handle this? After all, the definition of `LinearIndices` is
```julia
struct LinearIndices{N,R<:NTuple{N,AbstractUnitRange{Int}}} <: AbstractArray{Int,N}
    indices::R
end
```
and so some of the same tricks you're using here could be used there too. You might have to reset the integer at the end of each "column," perhaps by maintaining a hybrid linear/cartesian iterator:
```julia
julia> inds = LinearIndices((1:3, 1:4))
3×4 LinearIndices{2, Tuple{UnitRange{Int64}, UnitRange{Int64}}}:
 1  4  7  10
 2  5  8  11
 3  6  9  12

julia> inds[1:2:3,:]   # can't just increase by a constant
2×4 Matrix{Int64}:
 1  4  7  10
 3  6  9  12
```
Only, of course, when it's a StepRange.

Keep in mind that converting Cartesian to Linear is pretty fast (multiply and add); it's the opposite transition, Linear to Cartesian, that requires `div` and hence is dirt-slow. (That's why Cartesian is the fallback.)